### PR TITLE
fix(lsp): delete bufvar inside WinClosed event

### PR DESCRIPTION
**Pull request by glepnir**
Problem: floaing preview window can be closed by some ex commands like `only` `fclose` which will not clean the bufvar

Solution: use WinClosed event with floating_winnr for clean bufnr, and add test cases for vim.lsp.util.open_floating_preview

closes #31310

### CODE REVIEW

File: runtime/lua/vim/lsp/util.lua
Comment (mamekoro) [inline runtime/lua/vim/lsp/util.lua:1]: There's an unused assignment at line 1612: `vim.w[floating_winnr].buf_hold_win = bufnr`
`buf_hold_win` is no longer used.

Comment (glepnir) [reply runtime/lua/vim/lsp/util.lua:1]: yes :)

File: runtime/lua/vim/lsp/util.lua [hunk 1]
```diff
@@ -1622,6 +1617,20 @@ function M.open_floating_preview(contents, syntax, opts)
     api.nvim_buf_set_var(bufnr, 'lsp_floating_preview', floating_winnr)
   end
 
+  local augroup_name = ('closing_floating_preview_%d'):format(floating_winnr)
+  local ok =
+    pcall(api.nvim_get_autocmds, { gourp = augroup_name, pattern = tostring(floating_winnr) })
```
Comment (lewis6991) [inline runtime/lua/vim/lsp/util.lua:1622]:
```suggestion
    pcall(api.nvim_get_autocmds, { group = augroup_name, pattern = tostring(floating_winnr) })
```

**Review by lewis6991 — APPROVED** [review]

File: runtime/lua/vim/lsp/util.lua [hunk 2]
```diff
@@ -1622,6 +1617,22 @@ function M.open_floating_preview(contents, syntax, opts)
     api.nvim_buf_set_var(bufnr, 'lsp_floating_preview', floating_winnr)
   end
 
+  local augroup_name = ('closing_floating_preview_%d'):format(floating_winnr)
```
Comment (justinmk) [inline runtime/lua/vim/lsp/util.lua:1620]: internal augroups should use `nvim_` prefix.

File: runtime/lua/vim/lsp/util.lua [hunk 3]
```diff
@@ -1622,6 +1617,22 @@ function M.open_floating_preview(contents, syntax, opts)
...
+    pcall(api.nvim_get_autocmds, { group = augroup_name, pattern = tostring(floating_winnr) })
+  if not ok then
+    api.nvim_create_autocmd('WinClosed', {
+      group = api.nvim_create_augroup(augroup_name, {}),
+      pattern = tostring(floating_winnr),
+      callback = function()
+        if api.nvim_buf_is_valid(bufnr) then
+          vim.b[bufnr].lsp_floating_preview = nil
+        end
+        api.nvim_del_augroup_by_name(augroup_name)
```
Comment (justinmk) [inline runtime/lua/vim/lsp/util.lua:1631]: Why not just `return true` here? Then maybe the `augroup_name` isn't needed, also?

File: test/functional/plugin/lsp/utils_spec.lua [hunk 4]
```diff
@@ -265,6 +267,38 @@ describe('vim.lsp.util', function()
 
         eq(56, opts.height)
       end)
+
+      describe('vim.lsp.util.open_floating_preview', function()
```
Comment (justinmk) [inline test/functional/plugin/lsp/utils_spec.lua:271]: why is this in the `make_floating_popup_options .. with 59 lines above, 20 lines below` describe() block? https://github.com/neovim/neovim/blob/beec377e905baca73e772080c4f276c800ad2a40/test/functional/plugin/lsp/utils_spec.lua#L246

**Pull request merged**